### PR TITLE
Add arguments to "chezmoi unmanaged" and "chezmoi managed"

### PR DESCRIPTION
**Pull request by RuijieYu**
Fixes twpayne/chezmoi#1985.

Allow arguments on chezmoi managed and unmanaged.  Per recommendations on twpayne/chezmoi#1985, no additional flags are added to either commands, and behaviors are very similar to the respective commands without arguments (such as not recursing into directories).

``` console
$ chezmoi unmanaged ~/.ssh ~/.gnupg
(list unmanaged files under the arguments)
$ chezmoi managed ~/.config/git ~/scripts
(list managed files under the arguments)
```

Note that in consideration of duplicated arguments and outputs, sorting and deduplication is done on both the arguments and the outputs for unmanaged, and on the outputs for managed.  This helps with the following hypothetical situation (exaggerated to hopefully convey the idea more clearly):

``` console
$ chezmoi unmanaged ~/{,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,}
(extremely duplicated lines of output)
$ chezmoi unmanaged ~/.config/{,a,b,c,d,e,f,g,h,*}
(outputs for a..h are duplicated many times)
```

### CODE REVIEW

**Comment by RuijieYu** [comment]
I see the workflow errors, will address them later today.

**Comment by twpayne** [comment]
Thanks for making the changes, the [contributing guide](https://www.chezmoi.io/developer/contributing-changes/) should help you fix up the remaining ones.

**Comment by twpayne** [comment]
Note that the commit messages should be:

    feat: Make unmanaged command accept destination directory arguments

and

    feat: Make managed command accept destination directory arguments

chezmoi generates its release notes directly from the commit messages. These commit messages communicate more clearly to the user what effect the commits have.

**Comment by RuijieYu** [comment]
Thanks for the comments, I will address them within 24 hours.

**Comment by twpayne** [comment]
Thank you, but please, no rush! This is open source with voluntary contributions :)

**Comment by RuijieYu** [comment]
Thanks for pointing me to the contributing site!  I was a bit stumbled on getting the correct format as the `golangci-lint` test asked me to run `gofmt -s` on the `unmanagedcmd.go` file, which did not do anything.

The 24 hours was not meant for me to _rush_ the changes, but I would probably end up being caught up with other things in life and forget about this PR, had I not set a short time period for myself. :)

**Comment by twpayne** [comment]
Thanks for updating this PR so that the automated checks pass :)

This PR currently doesn't have any tests for the newly-added functionality. Could you add some to demonstrate that the new code behaves as expected and so that the functionality doesn't get accidentally broken in the future? I'd suggest adding extra lines to https://github.com/twpayne/chezmoi/blob/master/pkg/cmd/testdata/scripts/managed.txt and https://github.com/twpayne/chezmoi/blob/master/pkg/cmd/testdata/scripts/unmanaged.txt that demonstrate passing arguments to the `managed` and `unmanaged` commands.

**Comment by RuijieYu** [comment]
Thanks for the update! I didn't see the test scripts, but I think I understand them by now. The test scripts will probably be added by the end of next week, as I am freer next week than this one.

**Comment by RuijieYu** [comment]
Incidentally, I was looking around the test scripts (as I was a bit confused on the `$HOME` reference and the like) and came up with this change to the managed tests.  It seems to me that the result of `chezmoi managed .dir` should be equivalent to the result of `chezmoi managed | grep -P ^.dir`, effectively looking only under the directory `.dir` if it exists.

Here below is the change with which I came up -- I'll amend it to my commits soon-ish if we can agree that it makes sense and tests for the correct scenario.  Probably more tests are needed but this should be a good starting point.  The diff is made on `pkg/cmd/testdata/scripts/managed.txt` against the current `feat-upstream-1985` branch at fa1abe20.

``` diff
modified   pkg/cmd/testdata/scripts/managed.txt
@@ -24,6 +24,10 @@ cmp stdout golden/managed-symlinks
 chezmoi managed --exclude=files
 cmp stdout golden/managed-except-files

+# test chezmoi managed with arguments
+chezmoi managed $HOME${/}.dir
+cmp stdout golden/managed-with-args
+
 chhome home2/user

 # test that chezmoi managed does not evaluate templates
@@ -84,6 +88,11 @@ cmp stdout golden/managed2
 .symlink
 .template
 script
+-- golden/managed-with-args --
+.dir
+.dir/file
+.dir/subdir
+.dir/subdir/file
 -- home/user/.local/share/chezmoi/.chezmoiremove --
 .remove
 -- home2/user/.local/share/chezmoi/create_dot_create.tmpl --
```

**Comment by bradenhilton** [comment]
@RuijieYu I was assuming the output would be as it is in some of your examples in [#1985](https://github.com/twpayne/chezmoi/issues/1985) (don't output the directory given in the argument, only subdirectories and files):

> ```
> $ chezmoi managed ~/.gnupg ~/.ssh # the commands can accept multiple paths
> .gnupg/gpg.conf
> .ssh/config
> ```

just wanted to make sure I understood the intent.

How do we feel about these tests?

<details>
<summary><code>managed</code></summary>

```diff
diff --git a/pkg/cmd/testdata/scripts/managed.txt b/pkg/cmd/testdata/scripts/managed.txt
index eef0556f..f9066cc1 100644
--- a/pkg/cmd/testdata/scripts/managed.txt
+++ b/pkg/cmd/testdata/scripts/managed.txt
@@ -24,6 +24,14 @@ cmp stdout golden/managed-symlinks
 chezmoi managed --exclude=files
 cmp stdout golden/managed-except-files

+# test that chezmoi managed evaluates arguments
+chezmoi managed $HOME${/}.create $HOME${/}.dir
+cmp stdout golden/managed-with-args
+
+# test that chezmoi managed skips non-existent arguments
+chezmoi managed $HOME${/}.create $HOME${/}.dir $HOME${/}non-existent-dir
+cmp stdout golden/managed-with-args-missing
+
 chhome home2/user

 # test that chezmoi managed does not evaluate templates
@@ -78,6 +86,18 @@ cmp stdout golden/managed2
 .dir
 .dir/subdir
 .symlink
+-- golden/managed-with-args --
+.create
+.dir
+.dir/file
+.dir/subdir
+.dir/subdir/file
+-- golden/managed-with-non-existent-args --
+.create
+.dir
+.dir/file
+.dir/subdir
+.dir/subdir/file
 -- golden/managed2 --
 .create
 .file
```

</details>

<details>
<summary><code>unmanaged</code></summary>

```diff
diff --git a/pkg/cmd/testdata/scripts/unmanaged.txt b/pkg/cmd/testdata/scripts/unmanaged.txt
index dde48b07..d3ec9158 100644
--- a/pkg/cmd/testdata/scripts/unmanaged.txt
+++ b/pkg/cmd/testdata/scripts/unmanaged.txt
@@ -12,6 +12,13 @@ rm $CHEZMOISOURCEDIR/dot_file
 chezmoi unmanaged
 cmp stdout golden/unmanaged-dir-file

+chhome home2/user
+mkhomedir
+mksourcedir
+
+chezmoi unmanaged $HOME${/}managed-dir
+cmp stdout golden/unmanaged-dir-args
+
 -- golden/unmanaged --
 .local
 -- golden/unmanaged-dir --
@@ -21,3 +28,12 @@ cmp stdout golden/unmanaged-dir-file
 .dir
 .file
 .local
+-- golden/unmanaged-dir-args --
+managed-dir/unmanaged-dir
+managed-dir/unmanaged-dir2
+-- home2/user/.local/share/chezmoi/managed-dir/file --
+# contents of home2/user/managed-dir/file
+-- home2/user/managed-dir/unmanaged-dir/file --
+# contents of home2/user/managed-dir/unmanaged-dir/file
+-- home2/user/managed-dir/unmanaged-dir2/file --
+# contents of home2/user/managed-dir/unmanaged-dir2/file
```

</details>

**Comment by RuijieYu** [comment]
@twpayne and I discussed later in #1985, and concluded that we should stick to the current behavior of `chezmoi unmanaged`, where `chezmoi` should not descend inside an unmanaged directory.  I believe currently `chezmoi managed` already lists all individual files, so that should be fine.

The `managed-with-args` test for `managed` looks correct to me -- although I didn't take time  looking into how `create` works, so I can't confirm that with full confidence.

The remaining two tests looks fine.  I think some further test cases are needed to touch on its interoperability with other flags and some corner cases like symlinks, which I should be able to find time working on in the upcoming few days.

**Comment by RuijieYu** [comment]
I just updated the commits with additional tests -- please let me know if you think any more tests are needed.

There is one thing that I spent time to decide: testing the error message for a failure condition on `chezmoi unmanaged`.  There does not seem to be an easy way to match the "golden" file with the standard error that the test case was receiving (as there seems to be a rogue newline at the end).  I eventually figured that the specific error message on `chezmoi unmanaged $HOME${/}absent-path` should not matter, and all that matters is *that* chezmoi should fail with absent paths.

**Comment by bradenhilton** [comment]
Not a formal review, just a few things I was curious about:
(Of course we should wait for @twpayne's thoughts)

1. You might need to add something like `[*target*...]` to the headers in the doc pages:
    - ``# `managed` [*target*...]``
    - ``# `unmanaged` [*target*...]``
2. If you do the above, you might also want to add a sentence or two about what specifying an argument actually does.
3. The usage strings for other commands seem to use `[target(s)]...` conventionally. May or may not be needed here.
4. Can the commented out code be safely removed?

**Comment by RuijieYu** [comment]
The commented-out code was a remainder from when I was implementing it and forgot to take it out.  I will remember to take them out when I rebase next time.

Regarding your other point, I noticed that the `managed.md` file corresponds to [this page](https://www.chezmoi.io/reference/commands/managed/#-i-include-types) on the online documentation, so I think it is indeed helpful to add something into this file and `unmanaged.md` to reflect what the arguments do in either command.

**Comment by RuijieYu** [comment]
This is what I came up with.  The only thing I am currently debating about is whether we should somehow split the `*target*s` into its own subsection.

``` diff
modified   assets/chezmoi.io/docs/reference/commands/managed.md
@@ -1,6 +1,8 @@
-# `managed`
+# `managed` [*target*...]

-List all managed entries in the destination directory in alphabetical order.
+List all managed entries in the destination directory under all *target*s in
+alphabetical order.  When no *target*s are supplied, list all managed entries in
+the destination directory in alphabetical order.

 ## `-i`, `--include` *types*

modified   assets/chezmoi.io/docs/reference/commands/unmanaged.md
@@ -1,6 +1,9 @@
-# `unmanaged`
+# `unmanaged` [*target*...]

-List all unmanaged files in the destination directory.
+List all unmanaged files in *target*s.  When no *target*s are supplied, list all
+unmanaged files in the destination directory.
+
+It is an error to supply *target*s that are not found on the filesystem.

 !!! example
```

**Review by twpayne — APPROVED** [review]
Thank you @RuijieYu for the code and @bradenhilton for reviewing. This looks great!

@RuijieYu would you mind tidying up the debug comments? No worries if you don't have time to do this, I can do it in a follow-up commit.

**Comment by RuijieYu** [comment]
Thanks! I believe all my debug prints have been cleared up -- please let me know if I have missed any segments.

**Comment by twpayne** [comment]
Thank you very much for your work implementing this nice feature and your patience in getting it merged :)
